File: markdown/private/core/raw_version.py
import argparse
import json

from markdown.private.utils.metadata import Version

REPO_KEY = "STABLE_WORKSPACE_PARENT_REPO"
VERSION_KEY = "STABLE_WORKSPACE_PARENT_VERSION"


def from_version_file(file: str) -> Version:
    with open(file, encoding="utf-8") as f:
        return Version.model_validate_json(f.read())


def from_info_file(file: str) -> Version:
    version = ""
    repo = ""

    with open(file, encoding="utf-8") as f:
        for line in f:
            if line.startswith(REPO_KEY):
                repo = line[len(REPO_KEY) :].strip()
            elif line.startswith(VERSION_KEY):
                version = line[len(VERSION_KEY) :].strip()

    if not version or not repo:
        return Version(
            version="unversioned",
            repo="unversioned",
        )

    return Version(
        version=version,
        repo=repo,
    )


def main() -> None:
    parser = argparse.ArgumentParser()
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--version_file")
    group.add_argument("--info_file")
    parser.add_argument("outfile")
    args = parser.parse_args()

    if args.version_file:
        version = from_version_file(args.version_file)
    elif args.info_file:
        version = from_info_file(args.info_file)
    else:
        raise ValueError("Neither version file specified")

    with open(args.outfile, mode="w", encoding="utf-8") as f:
        json.dump(
            version.model_dump(
                mode="json",
                by_alias=True,
                exclude_unset=True,
                exclude_defaults=True,
            ),
            f,
            sort_keys=True,
            indent=4,
        )


if __name__ == "__main__":
    main()

File: markdown/private/core/raw_version.rs
use clap::{Args, Parser};
use markdown::metadata::Version;
use markdown::pretty_json::Write;
use std::error::Error;
use std::fs::read_to_string;
use std::io;

const REPO_KEY: &str = "STABLE_WORKSPACE_PARENT_REPO";
const VERSION_KEY: &str = "STABLE_WORKSPACE_PARENT_VERSION";

#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Cli {
    out_file: String,
    #[command(flatten)]
    in_file: InFile,
}

#[derive(Args)]
#[group(required = true, multiple = false)]
struct InFile {
    #[arg(long = "version_file")]
    version_file: Option<String>,
    #[arg(long = "info_file")]
    info_file: Option<String>,
}

fn from_version_file(path: &str) -> io::Result<Version> {
    let data = read_to_string(path)?;
    let version: Version = serde_json::from_str(&data)?;
    Ok(version)
}

fn from_info_file(path: &str) -> io::Result<Version> {
    let mut version = None;
    let mut repo = None;

    let data = read_to_string(path)?;
    for line in data.lines() {
        if let Some(line) = line.strip_prefix(REPO_KEY) {
            repo = Some(line.trim());
        } else if let Some(line) = line.strip_prefix(VERSION_KEY) {
            version = Some(line.trim())
        }
    }

    if let (Some(version), Some(repo)) = (version, repo) {
        return Ok(Version::new(version, repo));
    }

    Ok(Version::new("unversioned", "unversioned"))
}

fn main() -> Result<(), Box<dyn Error>> {
    let args = Cli::parse();

    let version = if let Some(path) = args.in_file.version_file {
        from_version_file(&path)?
    } else if let Some(path) = args.in_file.info_file {
        from_info_file(&path)?
    } else {
        return Err("Neither version file specified".into());
    };

    version.write(&args.out_file)
}
